### PR TITLE
Fix false positive validation result on form submission with AutovalidateMode.onUnfocus

### LINKED ISSUE #159378: False positive validation result on form submission using onUnfocus AutovalidateMode on Mobile devices
### Steps to reproduce

**Note**: This issue seems to be reproducible only on mobile devices. Tested on Android 13 (API 33) emulator.

1. Create a `Form` widget with `autovalidateMode` set to `AutovalidateMode.onUnfocus`
    - Add a `TextFormField` inside the form with a `validator` function that checks for validity (e.g., non-empty input)
2. Run the app and focus on the `TextFormField`
    - While the input is in an invalid state (e.g., leave the field empty or enter invalid data), keep the field focused without triggering any other events
3. Trigger a form submission to validate all fields in the form

### Expected results

- The validation logic for the focused `TextFormField` should execute and return an error message if the input is invalid
- The form's validation state should reflect the invalid state of the focused input

### Actual results

- The focused `TextFormField` is bypassed during form validation, and no error message is displayed for it
- The form incorrectly considers the focused input as valid, leading to unexpected behavior

**Workaround**:

To work around this issue, I had to explicitly unfocus the field and introduce a slight delay to force the validation logic to execute for the focused input:

```dart
onPressed: () async {
  FocusScope.of(context).unfocus(); // Unfocus the current field
  await Future<void>.delayed(Duration.zero); // Add a slight delay
  log('valid: ${key.currentState!.validate()}'); // Trigger validation
}
```

### Code sample

<details open><summary>Code sample</summary>

```dart
class MyForm extends StatefulWidget {
  const MyForm({super.key});

  @override
  State<MyForm> createState() => _MyFormState();
}

class _MyFormState extends State<MyForm> {
  final key = GlobalKey<FormState>();

  @override
  Widget build(BuildContext context) {
    return Form(
      key: key,
      autovalidateMode: AutovalidateMode.onUnfocus,
      child: Column(
        children: [
          TextFormField(
            validator: (String? value) => value == null || value.trim().isEmpty
                ? 'This field is required'
                : null,
          ),
          FilledButton(
            child: const Text('Submit'),
            onPressed: () {
              print('valid: ${key.currentState!.validate()}');
            },
          ),
        ],
      ),
    );
  }
}
```

</details>

### Screenshots or Video

_No response_

### Logs

_No response_

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.24.5, on macOS 14.6.1 23G93 darwin-x64, locale en-CO)
    • Flutter version 3.24.5 on channel stable at $HOME/Development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision dec2ee5c1f (9 days ago), 2024-11-13 11:13:06 -0800
    • Engine revision a18df97ca5
    • Dart version 3.5.4
    • DevTools version 2.37.3

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    • Android SDK at $HOME/Library/Android/sdk
    • Platform android-34, build-tools 33.0.2
    • Java binary at: $HOME/Library/Java/JavaVirtualMachines/corretto-17.0.13/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment Corretto-17.0.13.11.1 (build 17.0.13+11-LTS)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 16.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 16A242d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 21.0.3+-79915915-b509.11)

[✓] VS Code (version 1.95.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.100.0

[✓] Connected device (3 available)
    • sdk gphone64 x86 64 (mobile) • emulator-5554 • android-x64    • Android 13 (API 33) (emulator)
    • macOS (desktop)              • macos         • darwin-x64     • macOS 14.6.1 23G93 darwin-x64
    • Chrome (web)                 • chrome        • web-javascript • Google Chrome 131.0.6778.86

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by Mairramer**
Fixes #159378
Resolved an issue where focused TextFormField was bypassed during form validation
when using AutovalidateMode.onUnfocus on mobile devices. Validation is now correctly
triggered for all fields, including those that are focused.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/test/widgets/form_test.dart [hunk 1]
```diff
@@ -1497,4 +1497,54 @@ void main() {
...
+    // Focus on textFormField1.
+    await tester.tap(textFormField1);
+    await tester.pumpAndSettle();
+
+    // Check no error messages are displayed initially.
+    expect(find.text('foo/error'), findsNothing);
+    expect(find.text('bar/error'), findsNothing);
+
+    // Validate all fields manually using FormState.
+    expect(formKey.currentState!.validate(), isFalse); // Validation should return false due to errors.
```
Comment (nate-thegrate) [inline packages/flutter/test/widgets/form_test.dart:1543]: IMO the comment on the above line explains it nicely 👍

```suggestion
    expect(formKey.currentState!.validate(), isFalse);
```

File: packages/flutter/test/widgets/form_test.dart [hunk 2]
```diff
@@ -1497,4 +1497,54 @@ void main() {
...
     expect(find.text(errorText('bar')), findsOneWidget);
   });
+
+  testWidgets('AutovalidateMode.onUnfocus should validate all fields manually with FormState', (WidgetTester tester) async {
+    final GlobalKey<FormState> formKey = GlobalKey<FormState>();
+    String errorText(String? value) => '$value/error';
+
+    await tester.pumpWidget(
+      MaterialApp(
+        theme: ThemeData(),
```
Comment (nate-thegrate) [inline packages/flutter/test/widgets/form_test.dart:1507]: I don't believe theme data is necessary here:

```suggestion
```

File: packages/flutter/lib/src/widgets/form.dart [hunk 3]
```diff
@@ -354,9 +354,11 @@ class FormState extends State<Form> {
     final bool validateOnFocusChange = widget.autovalidateMode == AutovalidateMode.onUnfocus;
 
     for (final FormFieldState<dynamic> field in _fields) {
-      if (!validateOnFocusChange || !field._focusNode.hasFocus) {
+      final bool hasFocus = field._focusNode.hasFocus;
+
+      if (!validateOnFocusChange || !hasFocus || (validateOnFocusChange && hasFocus)) {
         final bool isFieldValid = field.validate();
-        hasError = !isFieldValid || hasError;
+        hasError |= !isFieldValid;
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/widgets/form.dart:361]: Not often that we see this kind of boolean logic!

File: packages/flutter/test/widgets/form_test.dart [hunk 4]
```diff
@@ -1497,4 +1497,54 @@ void main() {
...
+            ),
+          ),
+        ),
+      ),
+    );
+
+    final Finder textFormField1 = find.byKey(const Key('textFormField1'));
+
+    // Focus on textFormField1.
+    await tester.tap(textFormField1);
```
Comment (nate-thegrate) [inline packages/flutter/test/widgets/form_test.dart:1535]: Perhaps we could clean this up a tiny bit:

```dart
    // Focus on textFormField1.
    await tester.tap(find.byKey(const Key('textFormField1')));
```

alternatively:
```dart
    final GlobalKey<FormState> formKey = GlobalKey<FormState>();
    const Key fieldKey = Key('form field');

    // can be referenced in the pumpWidget call

    await tester.tap(find.byKey(fieldKey));
```

**Review by nate-thegrate — APPROVED** [review]
LGTM!

I can confirm the test fails when it runs against the current master branch.

Just have a few tiny suggestions :)

**Comment by nate-thegrate** [comment]
@Mairramer quick heads-up that we've got a merge conflict now.

File: packages/flutter/lib/src/widgets/form.dart (hunk 3 above)
Comment (justinmc) [reply packages/flutter/lib/src/widgets/form.dart:361]: Bonus points for `|=` for sure 👍 .

File: packages/flutter/test/widgets/form_test.dart [hunk 5]
```diff
@@ -1505,4 +1505,53 @@ void main() {
...
+    await tester.tap(find.byKey(fieldKey));
+    await tester.pumpAndSettle();
+
+    // Check no error messages are displayed initially.
+    expect(find.text('foo/error'), findsNothing);
+    expect(find.text('bar/error'), findsNothing);
+
+    // Validate all fields manually using FormState.
+    expect(formKey.currentState!.validate(), isFalse);
+    await tester.pumpAndSettle();
```
Comment (justinmc) [inline packages/flutter/test/widgets/form_test.dart:1551]: Could this just be `pump` or does that cause it to fail?

**Review by justinmc — APPROVED** [review]
LGTM, thanks for another form fix!

I verified that this is actually a bug over on the issue, see my comment: https://github.com/flutter/flutter/issues/159378#issuecomment-2533254278

File: packages/flutter/test/widgets/form_test.dart (hunk 5 above)
Comment (Mairramer) [reply packages/flutter/test/widgets/form_test.dart]: done!

**Pull request merged**